### ISSUE TITLE
Does it support the Transformers library?

### ISSUE THREAD

**Issue by kadirnar** [mobiusml/gemlite#27]
I want to use gemlite/hqq to speed up Orpheus-TTS training. However, I want to optimize in bfloat16 format. HQQ doesn't support this. I saw in a Twitter post that the gemlite library supports bfloat16. How can I implement this? 

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, Trainer, TrainingArguments, AutoTokenizer
import numpy as np
import yaml
import wandb

config_file = "config.yaml"

with open(config_file, "r") as file:
    config = yaml.safe_load(file)

dsn = config["TTS_dataset"]

model_name = config["model_name"]
run_name = config["run_name"]
project_name = config["project_name"]
base_repo_id = config["save_folder"]
epochs = config["epochs"]
batch_size = config["batch_size"]
save_steps = config["save_steps"]
pad_token = config["pad_token"]
number_processes = config["number_processes"]
learning_rate = config["learning_rate"]

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, attn_implementation="flash_attention_2")


ds = load_dataset(dsn, split="train") 

wandb.init(project=project_name, name = run_name)

training_args = TrainingArguments(
    overwrite_output_dir=True,
    num_train_epochs=epochs,
    per_device_train_batch_size=batch_size, 
    logging_steps=1,
    bf16=True,
    output_dir=f"./{base_repo_id}",
    report_to="wandb", 
    save_steps=save_steps,
    remove_unused_columns=True, 
    learning_rate=learning_rate,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=ds,
)

trainer.train()

**Comment by mobicham** [comment]
Hey, GemLite is for inference only, you can't use it for training.

In general, for inference, you just pass `HqqConfig` to your transformers model, and then you replace all the HQQLinear layers with GemLite:

``` Python
from transformers import AutoModelForCausalLM, HqqConfig 
quant_config = HqqConfig(nbits=4, group_size=64, axis=1, skip_modules=...)
model = AutoModelForCausalLM.from_pretrained(model_name, ..., quantization_config=quant_config)

from hqq.utils.patching import prepare_for_inference
prepare_for_inference(model, backend='gemlite') 
```

However, this will not work for your training case. 

You can still train with <a href='https://github.com/huggingface/peft/blob/fb7f2796e5411ee86588447947d1fdd5b6395cad/tests/test_gpu_examples.py#L2386C28-L2428">HQQ + LoRA without gemlite </a> though, make sure you use the backend to 
```Python
HQQLinear.set_backend(HQQBackend.PYTORCH_COMPILE)  
```

**Comment by kadirnar** [comment]
@mobicham  Can I use hqq in bfloat16 format for inference?What can I do with Gemlite bfloat16 support?

**Comment by mobicham** [comment]
Yes, it should work, just use ```...from_pretrained(model_id, ...torch_dtype=torch.bfloat16,...```

**Comment by mobicham** [comment]
@kadirnar it's not a problem on the gemlite side, I think it's probably some issue with the hooks in transformers. 
Unfortunately I don't have the time right now to look into it, but for the moment, you can load the model on cpu then quantize manually as you can see the example above, it should work fine.
